src/project-converter.ts

```diff
@@ -0,0 +1,159 @@
+  readonly conversionSettings: ConversionSettings;
+
+  /**
+   * A cache of all JS Exports by namespace, to map implicit HTML imports to
+   * explicit named JS imports.
+   */
+  readonly namespacedExports = new Map<string, JsExport>();
+  /**
+   * A cache of all converted documents. Document conversions should be
+   * idempotent, so conversion results can be safely cached.
+   */
+  readonly conversionResults = new Map<OriginalDocumentUrl, ConversionResult>();
+
+  constructor(
+      urlHandler: UrlHandlerInterface, conversionSettings: ConversionSettings) {
+    this.urlHandler = urlHandler;
+    this.conversionSettings = conversionSettings;
+  }
+
+  /**
+   * Convert a document and any of its dependencies. The output format (JS
+   * Module or HTML Document) is determined by whether the file is included in
+   * conversionSettings.includes.
+   */
+  convertDocument(document: Document) {
+    try {
+      this.conversionSettings.includes.has(document.url) ?
+          this.convertDocumentToJs(document, new Set()) :
+          this.convertDocumentToHtml(document, new Set());
+    } catch (e) {
+      console.error(`Error in ${document.url}`, e);
+    }
+  }
+
+  /**
+   * Check if a document is explicitly excluded or has already been converted
+   * to decide if it should be converted or skipped.
+   */
+  shouldConvertDocument(document: Document): boolean {
+    const documentUrl = getDocumentUrl(document);
+    if (this.conversionResults.has(documentUrl)) {
+      return false;
+    }
+    for (const exclude of this.conversionSettings.excludes) {
+      if (documentUrl.endsWith(exclude)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Specifically convert an HTML document to a JS module. Useful during
+   * conversion for dependencies where the type of result is explictly expected.
+   */
+  convertDocumentToJs(document: Document, visited: Set<OriginalDocumentUrl>) {
+    if (!this.shouldConvertDocument(document)) {
+      return;
+    }
+    const documentConverter = new DocumentConverter(this, document, visited);
+    const newModule = documentConverter.convertToJsModule();
+    this.handleConversionResult(newModule);
+  }
+
+  /**
+   * Specifically convert an HTML document without changing the file type
+   * (changes imports and inline scripts to modules as necessary.) Useful during
+   * conversion for dependencies where the type of result is explictly expected.
+   */
+  convertDocumentToHtml(document: Document, visited: Set<OriginalDocumentUrl>) {
+    if (!this.shouldConvertDocument(document)) {
+      return;
+    }
+    const documentConverter = new DocumentConverter(this, document, visited);
+    const newModule = documentConverter.convertAsToplevelHtmlDocument();
+    this.handleConversionResult(newModule);
+  }
+
+  /**
+   * A private instance method for handling new conversion results, exports,
+   * etc.
+   */
+  private handleConversionResult(newModule: ConversionResult): void {
```

Comment:
`_handleConversionResult`

Reply:
Some things in flight across multiple PRs should also be changed to `_`-prefix, adding a TODO to go back after everything is merged.